File: template/go/entra_sidecar_client.go
package main

import (
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"strings"
	"time"
)

type EntraSidecarClient struct {
	config AppConfig
	client *http.Client
}

func newEntraSidecarClient(config AppConfig) EntraSidecarClient {
	return EntraSidecarClient{
		config: config,
		client: &http.Client{Timeout: 30 * time.Second},
	}
}

func (c EntraSidecarClient) getAuthorizationHeader(incomingAuthorizationHeader string) (string, error) {
	mode := strings.ToLower(c.config.EntraSidecarAuthMode)
	if mode != "autonomous" && mode != "obo" {
		return "", fmt.Errorf("ENTRA_SIDECAR_AUTH_MODE must be autonomous or obo")
	}

	endpoint := "AuthorizationHeaderUnauthenticated"
	if mode == "obo" {
		endpoint = "AuthorizationHeader"
	}
	requestURL := fmt.Sprintf(
		"%s/%s/%s?AgentIdentity=%s&optionsOverride.RequestAppToken=%t",
		strings.TrimRight(c.config.EntraSidecarURL, "/"),
		endpoint,
		url.PathEscape(c.config.EntraSidecarServiceName),
		url.QueryEscape(c.config.EntraAgentClientID),
		mode == "autonomous",
	)
	request, err := http.NewRequest(http.MethodGet, requestURL, nil)
	if err != nil {
		return "", err
	}
	if mode == "obo" {
		if strings.TrimSpace(incomingAuthorizationHeader) == "" {
			return "", fmt.Errorf("the current request authorization header is required in obo mode")
		}
		request.Header.Set(
			"Authorization",
			normalizeAuthorizationHeader(incomingAuthorizationHeader),
		)
	}

	response, err := c.client.Do(request)
	if err != nil {
		return "", fmt.Errorf("Entra sidecar token request failed: %w", err)
	}
	defer response.Body.Close()
	body, err := io.ReadAll(response.Body)
	if err != nil {
		return "", err
	}
	if response.StatusCode < 200 || response.StatusCode >= 300 {
		return "", fmt.Errorf(
			"Entra sidecar token request failed: %d%s",
			response.StatusCode,
			formatResponseDetail(body),
		)
	}

	var payload struct {
		AuthorizationHeader string `json:"authorizationHeader"`
	}
	if err := json.Unmarshal(body, &payload); err != nil {
		return "", fmt.Errorf("invalid Entra sidecar response: %w", err)
	}
	header := normalizeAuthorizationHeader(payload.AuthorizationHeader)
	if header == "" {
		return "", fmt.Errorf("the Entra sidecar returned an empty authorizationHeader")
	}
	return header, nil
}

func normalizeAuthorizationHeader(value string) string {
	header := strings.TrimSpace(value)
	if header == "" || strings.HasPrefix(header, "Bearer ") || strings.HasPrefix(header, "PoP ") {
		return header
	}
	return "Bearer " + header
}

func formatResponseDetail(body []byte) string {
	detail := strings.TrimSpace(string(body))
	if detail == "" {
		return ""
	}
	return " - " + detail
}

File: template/rust/src/entra_sidecar_client.rs
use crate::config::AppConfig;
use serde_json::Value;

#[derive(Clone, Debug)]
pub struct EntraSidecarClient {
    config: AppConfig,
}

impl EntraSidecarClient {
    pub fn new(config: AppConfig) -> Self {
        Self { config }
    }

    pub fn get_authorization_header(
        &self,
        incoming_authorization_header: &str,
    ) -> Result<String, String> {
        let mode = self.config.entra_sidecar_auth_mode.to_ascii_lowercase();
        if mode != "autonomous" && mode != "obo" {
            return Err("ENTRA_SIDECAR_AUTH_MODE must be autonomous or obo.".to_string());
        }

        let endpoint = if mode == "autonomous" {
            "AuthorizationHeaderUnauthenticated"
        } else {
            "AuthorizationHeader"
        };
        let url = format!(
            "{}/{}/{}?AgentIdentity={}&optionsOverride.RequestAppToken={}",
            self.config.entra_sidecar_url.trim_end_matches('/'),
            endpoint,
            self.config.entra_sidecar_service_name,
            self.config.entra_agent_client_id,
            if mode == "autonomous" {
                "true"
            } else {
                "false"
            }
        );
        let mut request = minreq::get(url).with_timeout(30);

        if mode == "obo" {
            if incoming_authorization_header.trim().is_empty() {
                return Err(
                    "The current request authorization header is required in obo mode."
                        .to_string(),
                );
            }
            request = request.with_header(
                "Authorization",
                normalize_authorization_header(incoming_authorization_header)?,
            );
        }

        let response = request
            .send()
            .map_err(|error| format!("Entra sidecar token request failed: {error}"))?;
        let status = response.status_code;
        let body = response
            .as_str()
            .map_err(|error| format!("Unable to read Entra sidecar response: {error}"))?;
        if !(200..300).contains(&status) {
            let detail = if body.trim().is_empty() {
                String::new()
            } else {
                format!(" - {}", body.trim())
            };
            return Err(format!(
                "Entra sidecar token request failed: {}{}",
                status, detail
            ));
        }

        let payload: Value = serde_json::from_str(&body)
            .map_err(|error| format!("Invalid Entra sidecar response: {error}"))?;
        let header = payload["authorizationHeader"].as_str().ok_or_else(|| {
            "The Entra sidecar response did not include authorizationHeader.".to_string()
        })?;
        normalize_authorization_header(header)
    }
}

pub fn normalize_authorization_header(value: &str) -> Result<String, String> {
    let header = value.trim();
    if header.is_empty() {
        return Err("The authorization header is empty.".to_string());
    }
    if header.starts_with("Bearer ") || header.starts_with("PoP ") {
        Ok(header.to_string())
    } else {
        Ok(format!("Bearer {header}"))
    }
}
